File: src/main/java/com/jloom/orchestrate/ModuleApplier.java
package com.jloom.orchestrate;

import com.jloom.compose.RecipeComposer;
import com.jloom.exec.GradleRewriteRunner;
import com.jloom.registry.ModuleManifest;
import com.jloom.registry.ModuleRegistry;
import com.jloom.scaffold.FileTreeCopier;
import com.jloom.state.AppliedModule;
import com.jloom.state.ProjectState;
import com.jloom.state.ProjectStateStore;

import java.nio.file.Path;
import java.time.Instant;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class ModuleApplier {

    private final ModuleRegistry registry;
    private final RecipeComposer recipeComposer;
    private final GradleRewriteRunner runner;
    private final ProjectStateStore stateStore;
    private final FileTreeCopier fileCopier;

    public ModuleApplier(ModuleRegistry registry) {
        this(registry, new RecipeComposer(), new GradleRewriteRunner(),
                new ProjectStateStore(), new FileTreeCopier());
    }

    public ModuleApplier(ModuleRegistry registry, RecipeComposer recipeComposer,
                         GradleRewriteRunner runner, ProjectStateStore stateStore,
                         FileTreeCopier fileCopier) {
        this.registry = registry;
        this.recipeComposer = recipeComposer;
        this.runner = runner;
        this.stateStore = stateStore;
        this.fileCopier = fileCopier;
    }

    public ApplyResult apply(Path targetProject, List<String> moduleIds, Map<String, String> overrides,
                              boolean dryRun, String basePackage, String projectName) {
        ProjectState state = stateStore.load(targetProject);

        List<String> problems = registry.validate(state.appliedIds(), moduleIds);
        if (!problems.isEmpty()) {
            return ApplyResult.rejected(problems);
        }

        ProjectState tokenState = state.appliedIds().isEmpty() && (basePackage != null || projectName != null)
                ? state.withProjectName(projectName != null ? projectName : state.projectName())
                        .withBasePackage(basePackage != null ? basePackage : state.basePackage())
                : state;
        Map<String, String> projectTokens = projectLevelTokens(tokenState);

        List<RecipeComposer.ModuleSelection> selections = new ArrayList<>();
        Map<String, Map<String, String>> answersByModule = new LinkedHashMap<>();

        for (String id : moduleIds) {
            ModuleManifest manifest = registry.require(id);
            Map<String, String> answers = resolveDefaultsOnly(manifest, overrides);
            answersByModule.put(id, answers);

            if (!manifest.mergeRecipes().isEmpty()) {
                selections.add(new RecipeComposer.ModuleSelection(manifest, answers));
            }
        }

        if (!dryRun) {
            for (String id : moduleIds) {
                ModuleManifest manifest = registry.require(id);
                if (manifest.scaffold()) {
                    Map<String, String> fileTokens = new LinkedHashMap<>(projectTokens);
                    fileTokens.putAll(answersByModule.get(id));
                    fileCopier.copy(manifest, targetProject, fileTokens);
                }
            }
        }

        String recipeOutput = "no merges required";
        if (!selections.isEmpty()) {
            if (dryRun && !runner.hasGradleWrapper(targetProject)) {
                return finish(targetProject, state, moduleIds, answersByModule, dryRun,
                        "recipe preview skipped — no Gradle wrapper on disk yet to diff against",
                        basePackage, projectName);
            }

            String recipeYaml = recipeComposer.compose(selections);
            GradleRewriteRunner.Result result = runner.run(targetProject, recipeYaml, recipeComposer.recipeName(), dryRun);

            if (!result.success()) {
                return ApplyResult.failed(result.output());
            }
            recipeOutput = result.streamed() ? "" : result.output();
        }

        if (!dryRun) {
            for (String id : moduleIds) {
                ModuleManifest manifest = registry.require(id);
                if (!manifest.scaffold() && !manifest.fileTemplates().isEmpty()) {
                    Map<String, String> fileTokens = new LinkedHashMap<>(projectTokens);
                    fileTokens.putAll(answersByModule.get(id));
                    fileCopier.copy(manifest, targetProject, fileTokens);
                }
            }
        }

        return finish(targetProject, state, moduleIds, answersByModule, dryRun, recipeOutput, basePackage, projectName);
    }

    private static Map<String, String> resolveDefaultsOnly(ModuleManifest module, Map<String, String> overrides) {
        Map<String, String> answers = new LinkedHashMap<>();
        for (ModuleManifest.Prompt prompt : module.prompts()) {
            String overrideKey = module.id() + "." + prompt.key();
            if (overrides != null && overrides.containsKey(overrideKey)) {
                answers.put(prompt.key(), overrides.get(overrideKey));
                continue;
            }
            if (prompt.defaultValue() != null) {
                answers.put(prompt.key(), prompt.defaultValue());
                continue;
            }
            throw new IllegalArgumentException(
                    "Module '" + module.id() + "' requires --set " + overrideKey + "=<value> (no default)");
        }
        return answers;
    }

    private Map<String, String> projectLevelTokens(ProjectState state) {
        if (state.basePackage() == null && state.projectName() == null) {
            return Map.of();
        }
        Map<String, String> tokens = new LinkedHashMap<>();
        if (state.projectName() != null) {
            tokens.put("project_name", state.projectName());
        }
        if (state.basePackage() != null) {
            tokens.put("package", state.basePackage());
            tokens.put("package_path", state.basePackage().replace('.', '/'));
        }
        return tokens;
    }

    private ApplyResult finish(Path targetProject, ProjectState state, List<String> moduleIds,
                                Map<String, Map<String, String>> answersByModule, boolean dryRun, String output,
                                String basePackage, String projectName) {
        if (dryRun) {
            return ApplyResult.dryRun(output);
        }
        ProjectState seeded = state;
        if (state.appliedIds().isEmpty()) {
            String name = projectName != null ? projectName : targetProject.getFileName().toString();
            seeded = seeded.withProjectName(name);
            if (basePackage != null) {
                seeded = seeded.withBasePackage(basePackage);
            }
        }
        ProjectState updated = seeded;
        for (String id : moduleIds) {
            ModuleManifest manifest = registry.require(id);
            updated = updated.withApplied(new AppliedModule(
                    id, manifest.version(), Instant.now(), answersByModule.getOrDefault(id, Map.of())));
        }
        stateStore.save(targetProject, updated);
        return ApplyResult.applied(output);
    }

    public sealed interface ApplyResult {
        record Applied(String output) implements ApplyResult {
        }
        record DryRun(String diff) implements ApplyResult {
        }
        record Rejected(List<String> problems) implements ApplyResult {
        }
        record Failed(String output) implements ApplyResult {
        }

        static ApplyResult applied(String output) { return new Applied(output); }
        static ApplyResult dryRun(String diff) { return new DryRun(diff); }
        static ApplyResult rejected(List<String> problems) { return new Rejected(problems); }
        static ApplyResult failed(String output) { return new Failed(output); }
    }
}

File: bun/src/orchestrate/moduleApplier.ts
import path from "node:path";
import { applyOperations } from "../merge/executor.ts";
import { compose } from "../merge/recipeComposer.ts";
import type { ModuleSelection } from "../merge/recipeComposer.ts";
import { ModuleRegistry } from "../registry/moduleRegistry.ts";
import type { ModuleManifest } from "../registry/types.ts";
import { copy } from "../scaffold/fileTreeCopier.ts";
import {
  appliedIds,
  loadState,
  saveState,
  withApplied,
  withBasePackage,
  withProjectName,
} from "../state/projectStateStore.ts";
import type { ProjectState } from "../state/projectStateStore.ts";

export type ApplyResult =
  | { kind: "applied"; output: string }
  | { kind: "dryRun"; diff: string }
  | { kind: "rejected"; problems: string[] }
  | { kind: "failed"; output: string };

function resolveDefaultsOnly(module: ModuleManifest, overrides: Record<string, string>): Record<string, string> {
  const answers: Record<string, string> = {};
  for (const prompt of module.prompts) {
    const overrideKey = `${module.id}.${prompt.key}`;
    if (overrideKey in overrides) {
      answers[prompt.key] = overrides[overrideKey]!;
      continue;
    }
    if (prompt.defaultValue !== undefined) {
      answers[prompt.key] = prompt.defaultValue;
      continue;
    }
    throw new Error(`Module '${module.id}' requires --set ${overrideKey}=<value> (no default)`);
  }
  return answers;
}

function computeTokenState(
  state: ProjectState,
  basePackage: string | undefined,
  projectName: string | undefined,
): ProjectState {
  const isFirstApply = appliedIds(state).length === 0;
  if (!isFirstApply || (basePackage === undefined && projectName === undefined)) {
    return state;
  }
  const withName = withProjectName(state, projectName !== undefined ? projectName : state.projectName);
  return withBasePackage(withName, basePackage !== undefined ? basePackage : state.basePackage);
}

function projectLevelTokens(state: ProjectState): Record<string, string> {
  if (state.basePackage === undefined && state.projectName === undefined) {
    return {};
  }
  const tokens: Record<string, string> = {};
  if (state.projectName !== undefined) {
    tokens.project_name = state.projectName;
  }
  if (state.basePackage !== undefined) {
    tokens.package = state.basePackage;
    tokens.package_path = state.basePackage.split(".").join("/");
  }
  return tokens;
}

function finish(
  registry: ModuleRegistry,
  targetProject: string,
  state: ProjectState,
  moduleIds: string[],
  answersByModule: Map<string, Record<string, string>>,
  dryRun: boolean,
  output: string,
  basePackage: string | undefined,
  projectName: string | undefined,
): ApplyResult {
  if (dryRun) {
    return { kind: "dryRun", diff: output };
  }

  let seeded = state;
  if (appliedIds(state).length === 0) {
    const name = projectName !== undefined ? projectName : path.basename(targetProject);
    seeded = withProjectName(seeded, name);
    if (basePackage !== undefined) {
      seeded = withBasePackage(seeded, basePackage);
    }
  }

  let updated = seeded;
  for (const id of moduleIds) {
    const manifest = registry.require(id);
    updated = withApplied(updated, {
      id,
      version: manifest.version,
      appliedAt: new Date().toISOString(),
      answers: answersByModule.get(id) ?? {},
    });
  }
  saveState(targetProject, updated);
  return { kind: "applied", output };
}

/** Mirrors ModuleApplier.apply(...). Note on dry-run: the Java version could still shell out
 * to a real `gradlew rewriteDryRun` when a wrapper existed on disk, but its own CLI callers
 * never actually displayed that diff text to the user (always a static "Dry run — ..."
 * message regardless) - so dropping OpenRewrite loses no user-visible behavior here. This
 * version still calls compose(...) during a dry run (cheap, in-memory) so a malformed merge
 * fragment is still caught early, but never touches disk when dryRun is true. */
export function apply(
  registry: ModuleRegistry,
  targetProject: string,
  moduleIds: string[],
  overrides: Record<string, string>,
  dryRun: boolean,
  basePackage: string | undefined,
  projectName: string | undefined,
): ApplyResult {
  const state = loadState(targetProject);

  const problems = registry.validate(appliedIds(state), moduleIds);
  if (problems.length > 0) {
    return { kind: "rejected", problems };
  }

  const tokenState = computeTokenState(state, basePackage, projectName);
  const projectTokens = projectLevelTokens(tokenState);

  const selections: ModuleSelection[] = [];
  const answersByModule = new Map<string, Record<string, string>>();

  for (const id of moduleIds) {
    const manifest = registry.require(id);
    const answers = resolveDefaultsOnly(manifest, overrides);
    answersByModule.set(id, answers);
    if (manifest.mergeRecipes.length > 0) {
      selections.push({ manifest, answers });
    }
  }

  if (!dryRun) {
    for (const id of moduleIds) {
      const manifest = registry.require(id);
      if (manifest.scaffold) {
        const fileTokens = { ...projectTokens, ...answersByModule.get(id) };
        copy(manifest, targetProject, fileTokens);
      }
    }
  }

  let recipeOutput = "no merges required";
  if (selections.length > 0) {
    const operations = compose(selections);
    if (dryRun) {
      return finish(registry, targetProject, state, moduleIds, answersByModule, dryRun, "Dry run — no changes written.", basePackage, projectName);
    }
    try {
      applyOperations(targetProject, operations);
    } catch (err) {
      return { kind: "failed", output: err instanceof Error ? err.message : String(err) };
    }
    recipeOutput = `Applied ${operations.length} merge operation(s).`;
  }

  if (!dryRun) {
    for (const id of moduleIds) {
      const manifest = registry.require(id);
      if (!manifest.scaffold && manifest.fileTemplates.length > 0) {
        const fileTokens = { ...projectTokens, ...answersByModule.get(id) };
        copy(manifest, targetProject, fileTokens);
      }
    }
  }

  return finish(registry, targetProject, state, moduleIds, answersByModule, dryRun, recipeOutput, basePackage, projectName);
}
